src/utils_latency_config.c

```diff
@@ -0,0 +1,163 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+ * FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+ * DEALINGS IN THE SOFTWARE.
+ *
+ * Authors:
+ *   Florian octo Forster <octo at collectd.org>
+ *   Pavel Rochnyack <pavel2000 at ngs.ru>
+ */
+
+#include "collectd.h"
+#include "common.h"
+#include "utils_latency_config.h"
+
+int latency_config_add_percentile (const char *plugin, latency_config_t *cl,
+    oconfig_item_t *ci)
+{
+  if ((ci->values_num != 1) || (ci->values[0].type != OCONFIG_TYPE_NUMBER))
+  {
+    ERROR ("%s plugin: \"%s\" requires exactly one numeric argument.",
+             plugin, ci->key);
+    return (-1);
+  }
+
+  double percent = ci->values[0].value.number;
+  double *tmp;
+
+  if ((percent <= 0.0) || (percent >= 100))
+  {
+    ERROR ("%s plugin: The value for \"%s\" must be between 0 and 100, "
+        "exclusively.", plugin, ci->key);
+    return (ERANGE);
+  }
+
+  tmp = realloc (cl->percentile,
+      sizeof (*cl->percentile) * (cl->percentile_num + 1));
+  if (tmp == NULL)
+  {
+    ERROR ("%s plugin: realloc failed.", plugin);
+    return (ENOMEM);
+  }
+  cl->percentile = tmp;
+  cl->percentile[cl->percentile_num] = percent;
+  cl->percentile_num++;
+
+  return (0);
+} /* int latency_config_add_percentile */
+
+int latency_config_add_rate (const char *plugin, latency_config_t *cl,
+    oconfig_item_t *ci)
+{
+
+  if ((ci->values_num != 2)
+    ||(ci->values[0].type != OCONFIG_TYPE_NUMBER)
+    ||(ci->values[1].type != OCONFIG_TYPE_NUMBER))
+  {
+    ERROR ("%s plugin: \"%s\" requires exactly two numeric arguments.",
+             plugin, ci->key);
+    return (-1);
+  }
+
+  if (ci->values[1].value.number &&
+      ci->values[1].value.number <= ci->values[0].value.number)
+  {
+    ERROR ("%s plugin: MIN must be less than MAX in \"%s\".",
+             plugin, ci->key);
+    return (-1);
+  }
+
+  if (ci->values[0].value.number < 0.001)
+  {
+    ERROR ("%s plugin: MIN must be greater or equal to 0.001 in \"%s\".",
+             plugin, ci->key);
+    return (-1);
+  }
+
+  cdtime_t lower = DOUBLE_TO_CDTIME_T(ci->values[0].value.number);
+  cdtime_t upper = DOUBLE_TO_CDTIME_T(ci->values[1].value.number);
+  cdtime_t *tmp;
+
+  tmp = realloc (cl->rates,
+      sizeof (*cl->rates) * (cl->rates_num + 1) * 2);
+  if (tmp == NULL)
+  {
+    ERROR ("%s plugin: realloc failed.", plugin);
+    return (ENOMEM);
+  }
+  cl->rates = tmp;
+  cl->rates[cl->rates_num * 2] = lower;
+  cl->rates[cl->rates_num * 2 + 1] = upper;
+  cl->rates_num++;
+
+  return (0);
+} /* int latency_config_add_rate */
+
+
+int latency_config_copy (latency_config_t *dst, const latency_config_t src)
+{
+  /* Copy percentiles configuration */
+  dst->percentile_num = src.percentile_num;
+  dst->percentile = malloc(sizeof (*dst->percentile) * (src.percentile_num));
+  if (dst->percentile == NULL)
+    goto nomem;
```

Comment:
Please avoid using `goto`. Instead, use `latency_config_free()` for easy freeing of all fields and return:

```c
*dst = (latency_config_t) {
  .rates = NULL,
  .rates_num = src.rates_num,
  .rates_type = NULL,
  .percentile = NULL,
  .percentile_num = src.percentile_num,
  .percentile_type = NULL,
};
/* … */
if (dst->percentile == NULL)
{
  latency_config_free (*dst);
  return (ENOMEM);
}
```

Here and below.